qemu/tests/trim_support_test.py

```diff
@@ -0,0 +1,101 @@
+import logging
+import os
+
+from avocado.utils import process
+from virttest import utils_test
+from virttest import error_context
+from virttest import data_dir
+from virttest import utils_disk
+from virttest import utils_misc
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Test disk trimming in windows guest
+    1) boot the vm with a data disk
+    2) format the data disk without quick mode
+    3) check the disk file size in host, and record for compare
+    4) trim the data disk in guest
+    5) check the disk file again in host, the file size should shrink
+
+    :param test: QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env: Dictionary with test environment.
+    """
+
+    def _get_size_value(size_str):
+        """
+        Get size value form size string
+
+        :param size_str: data size string
+        :return: the size numeric value, measured by MB
+        """
+        size_str = utils_misc.normalize_data_size(size_str)
+        size = float(size_str)
+        return size
+
+    def _disk_size_smaller(ori_size):
+        """
+        Check till the disk size becomes smaller than ori_size
+        :param ori_size: original size to compare to
+        :return: new size if it smaller than ori_size, else None
+        """
+        output = process.system_output(host_check_cmd, shell=True).decode()
+        new_size = _get_size_value(str(output))
+        logging.info("Current data disk size: %sMB" % new_size)
+        if new_size < ori_size:
+            return new_size
+        return None
+
+    host_check_cmd = params.get("host_check_cmd")
+    image_dir = os.path.join(data_dir.get_data_dir(), 'images')
+    host_check_cmd %= image_dir
+    image_name = params["stg_name"]
+    stg_param = params.object_params(image_name)
+    image_size_str = stg_param["image_size"]
+    guest_trim_cmd = params["guest_trim_cmd"]
+    driver_name = params.get("driver_name", "netkvm")
+
+    timeout = float(params.get("timeout", 360))
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    session = vm.wait_for_login(timeout=timeout)
+    error_context.context("Check if the driver is installed and verified",
+                          logging.info)
+    session = utils_test.qemu.windrv_check_running_verifier(
+        session, vm, test, driver_name, timeout)
+
+    error_context.context("Format data disk", logging.info)
+    disk_index = utils_misc.wait_for(
+        lambda: utils_disk.get_windows_disks_index(session, image_size_str),
+        120)
+    if not disk_index:
+        test.error("Failed to get the disk index of size %s" % image_size_str)
+    if not utils_disk.update_windows_disk_attributes(session, disk_index):
+        test.error("Failed to bring data disk online")
```

Comment:
when disk is readonly or it's offline,  above function return false,  I suggest  change message to, 'fail to enable disk %s' % disk_index.